File: Source/NextTurnRuntime/Private/Slate/UICommandList.cpp
// Copyright (c) NextTurn. All rights reserved.
// Licensed under the Apache License, Version 2.0.
// See LICENSE.txt in the project root for more information.

#if WITH_APPLICATION_CORE

#include "NextTurnRuntime/Private/Exports.hpp"

#include "NextTurnRuntime/Private/Construct.hpp"
#include "NextTurnRuntime/Private/ManagedDelegate.hpp"
#include "NextTurnRuntime/Public/Compiler.hpp"

#include "Runtime/Core/Public/CoreMinimal.h"

#include "Runtime/Slate/Public/Framework/Commands/UICommandList.h"

ENABLE_WARNINGS;

#define TYPE_EXPORT_MANAGED_TYPE_NAME UICommandList

TYPE_EXPORT_START

void EXPORT_CALL_CONV Initialize(TSharedRef<FUICommandList>& CommandList)
{
	Construct(CommandList, TSharedRef<FUICommandList>{ new FUICommandList{} });
}

void EXPORT_CALL_CONV MapAction(
	const TSharedRef<FUICommandList>& CommandList,
	const TSharedRef<const FUICommandInfo>& CommandInfo,
	FExecuteAction::FStaticDelegate::FFuncPtr Execute,
	intptr_t ExecuteHandle)
{
	CommandList->MapAction(CommandInfo, TManagedDelegate<FExecuteAction>::Create(Execute, ExecuteHandle));
}

const FMemberSymbol Members[] =
{
	EXPORT_METHOD(Initialize),
	EXPORT_METHOD(MapAction),
};

TYPE_EXPORT_END

#undef TYPE_EXPORT_MANAGED_TYPE_NAME

RESTORE_WARNINGS;

#endif

File: Managed/NextTurn.UE.Runtime/Slate/UICommandList.cs
// Copyright (c) NextTurn. All rights reserved.
// Licensed under the Apache License, Version 2.0.
// See LICENSE.txt in the project root for more information.

using System;
using System.Runtime.InteropServices;
using NextTurn.UE.Annotations;

namespace Unreal.Slate
{
    public sealed class UICommandList : IDisposable
    {
        internal readonly SharedReference Reference;

        private bool disposed;

        public UICommandList() => NativeMethods.Initialize(out this.Reference);

        ~UICommandList() => this.DisposeImpl();

        public void Dispose()
        {
            this.DisposeImpl();
            GC.SuppressFinalize(this);
        }

        private void DisposeImpl()
        {
            if (!this.disposed)
            {
                this.Reference.ReleaseReference();

                this.disposed = true;
            }
        }

        public void MapAction(UICommandInfo commandInfo, Action execute) =>
            NativeMethods.MapAction(
                this.Reference,
                commandInfo.Reference,
                Marshal.GetFunctionPointerForDelegate(execute),
                GCHandle.ToIntPtr(GCHandle.Alloc(execute)));

        [CLSCompliant(false)]
        public unsafe void MapAction(UICommandInfo commandInfo, delegate* unmanaged<void> execute) =>
            NativeMethods.MapAction_(
                this.Reference,
                commandInfo.Reference,
                execute);

        private static class NativeMethods
        {
            [Calli]
            public static extern void Initialize(out SharedReference commandList);

            [Calli]
            public static extern void MapAction(
                in SharedReference commandList,
                in SharedReference commandInfo,
                IntPtr execute,
                IntPtr executeHandle);

            [Calli]
            public static extern unsafe void MapAction_(
                in SharedReference commandList,
                in SharedReference commandInfo,
                delegate* unmanaged<void> execute);
        }
    }
}
